Param mtlarray

christiangnrd opened the pull request:
My initial attempt at #190.

Only touches MtlArray because I could not figure out how to parameterize Obj-C objects. If it's possible, let me know and I'll parameterize MTLBuffer.

Will add docs once the code is reviewed.

Depends on #193

maleadt commented on src/array.jl:
```diff
@@ -73,34 +74,44 @@ mutable struct MtlArray{T,N} <: AbstractGPUArray{T,N}
-Base.similar(a::MtlArray{T}, dims::Base.Dims{N}) where {T,N} = MtlArray{T,N}(undef, dims; storage=storagemode(a))
-Base.similar(a::MtlArray, ::Type{T}, dims::Base.Dims{N}) where {T,N} =
-  MtlArray{T,N}(undef, dims; storage=storagemode(a))
+Base.similar(a::MtlArray{T,N,S}) where {T,N,S} =
+  MtlArray{T,N,S}(undef, size(a))
+Base.similar(a::MtlArray{T,<:Any,S}, dims::Base.Dims{N}) where {T,N,S} =
+  MtlArray{T,N,S}(undef, dims)
+Base.similar(a::MtlArray{<:Any,<:Any,S}, ::Type{T}, dims::Base.Dims{N}) where {T,N,S} =
+  MtlArray{T,N,S}(undef, dims)
```
A `similar` constructor to modify the storage is probably useful, e.g., `similar(mtl([1]); storage=Shared)`.

christiangnrd replied:
Done

maleadt commented on src/pool.jl:
```diff
@@ -31,33 +31,34 @@ Base.:(-)(a::AllocStats, b::AllocStats) = (;
-Note that `Private` buffers can't be directly accessed from the CPU, therefore you cannot
-use this option if you pass a ptr to initialize the memory.
-"""
+# """
+#     alloc(device, bytesize, [ptr=nothing];
+#           storage=Default, hazard_tracking=Default, cache_mode=Default)
+
+# Allocates a Metal buffer on `device` of`bytesize` bytes. If a CPU-pointer is passed as last
+# argument, then the buffer is initialized with the content of the memory starting at `ptr`,
+# otherwise it's zero-initialized.
+
+# ! Note: You are responsible for freeing the returned buffer
+
+# The storage kwarg controls where the buffer is stored. Possible values are:
+#  - Private : Residing on the device
+#  - Shared  : Residing on the host
+#  - Managed : Keeps two copies of the buffer, on device and on host. Explicit calls must be
+#    given to syncronize the two
+#  - Memoryless : an iOS specific thing that won't work on Mac.
+
+# Note that `Private` buffers can't be directly accessed from the CPU, therefore you cannot
+# use this option if you pass a ptr to initialize the memory.
+# """
```
Unintended change?

christiangnrd replied:
Yes good catch

maleadt commented:
Looks good, thanks!

> Only touches MtlArray because I could not figure out how to parameterize Obj-C objects. If it's possible, let me know and I'll parameterize MTLBuffer.

Which ObjC type did you want to parameterize?

christiangnrd commented:
I have addressed your review comments. I'll be adding documentation over the weekend.

> Which ObjC type did you want to parameterize?

I was thinking MTLbuffer. Would there be any benefits to doing so?

christiangnrd commented:
I added documentation similar to what CUDA.jl has except that I mention the `storage` parameter in `mtl` and `MtlArray`.

maleadt commented on test/array.jl:
```diff
@@ -23,75 +47,87 @@ end
 
-    # type and dimensionality specified, accepting dims as series of Ints
-    let arr = MtlArray{Int,3}(undef, dim[1],dim[2],dim[3]; storage=SM)
+    # xtype and dimensionality specified, accepting dims as series of Ints
```
```suggestion
    # type and dimensionality specified, accepting dims as series of Ints
```

maleadt commented on test/array.jl:
```diff
@@ -1,20 +1,44 @@
 STORAGEMODES = [Private, Shared]#, Managed]
 @testset "arrays ($SM)" for SM in STORAGEMODES
```
I'm not sure it's worth re-executing all tests for both storage modes. Maybe we should confine the tests that set the storage mode typevar to those that e.g. call `check_storagemode`, or (when we have those) do some storage mode-specific operations. Other tests could just leave the storage mode unspecified. WDYT?

christiangnrd replied:
I changed the tests to remove duplication. I also found some blind spots in the tests so I added a few. In total the number of tests in array went down very slightly.

maleadt reviewed (commented):
LGTM. Anything else you want to do here, since you hadn't marked the PR ready for review?

christiangnrd commented:
I forgot to mark it ready for review. Should be good to merge if you deem it so!
